### ISSUE TITLE
[BL] To refactor BL in order to work with different transceivers and apply retries

### ISSUE THREAD

**Issue by ralcaide** [Sinapse-Energia/IoT-Hub-Home#38]
It is necessary to refactor the BL code in order to work with the most updated transceiver classes allowing retries if needed.

The work done here will be reused in other products like M2M and CMC

The variables sharing with CL using flash should be also improved.

**Comment by juanra-bascones** [comment]
I have had to solve some preliminary problems before being able to effectively work in this branch.

1) **I needed to made a change in the project properties** in order to be able to use the debugger. 
As-downloaded, there wasn´t possible to use step by step debug execution (the flow was jumping back and forth in an apparent non-sense way).
Unable finding the cause by trial-error, I had to resort to line by line compare .cproject file with other fine working projects in order to identify the ofending parts, that finally happen to be related with optimization options.
I don´t know about the reason to have these optimization switches in the server code (and because of this I don´t plan to upload these changes by now) but from my point of view these flags should be better only used from release makefiles 

2) **I had to make a small change in the MX_GPIO_Init() function**
This initialization function seem to be specific for the Hub_Home board (in Deinitions.h defines DEFAULTBOARD to Tfiberboard, while this value is nor defined nor never used).
However, for some reason pending to be clarified the last Pin configuration call in the funcion, the one looking as 

```

	/*Configure GPIO pins : PAPin PAPin */
	GPIO_InitStruct.Pin = LED_RED_Pin | LED_GREEN_Pin;
	GPIO_InitStruct.Mode = GPIO_MODE_OUTPUT_PP;
	GPIO_InitStruct.Pull = GPIO_NOPULL;
	GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_LOW;
	HAL_GPIO_Init(GPIOA, &GPIO_InitStruct);


```
cause systematically the debuger to report a ST-Link remote error and to finish abnormally.
I needed to comment out this section in order to be able to go ahead with the issue goal
The cause for this failure needs to be investigated in paralel.

After those two previous changes, I'm now already working on the roadmap outlined in (https://github.com/Sinapse-Energia/bootloader-stm32/issues/43#issuecomment-508373740), that is fully valid for this product Bootloader too

**Comment by ralcaide** [comment]
Try to enable DMA for HLK/ETH, please. Feedback from Francis here:

https://github.com/Sinapse-Energia/M2M-CommBoard/issues/130

**Comment by juanra-bascones** [comment]
I'm quite confused with something that it seems new to me....
(At least, I don't remember it to happen when working on this with other boards).

In brief, what I am seeing now is that as trying to put the modem in **"command mode"** (what is done by applying a short low pulse to the ES0 Pin, as the User manul states in the page 10:
_ES0 Function:
1. After the system starts properly, ES0 left low 0.05s <t <6s, Serial port 0 to exit transparent transmission mode._
)

What It happens is that the modem, in fact, enters in command mode, but at the same time, **it loses all the stored configuration paramenters, like if it the factory settings would be restored.**

I don´t remember this to happen when I was working with Hub and presence boards. I will check it as sson as possible.

One thing I have noticed (I don´t know if it's related with it or not) is that the Transceiver in the Home Hub board I'm working with is a later model that the one in the other boards.
I have noticed that in the Network mode drop-down list box, it opens one more option that it used to do (one sixth option labeled "bridge").
Also, when pressing the Web Configuration page "Apply button", there is a feedback gadget to show the progress it wasn´t there before...

Because of this, I can suspect  that maybe  **the device I'm working with is not exactly the same HLK-RM08K as it were until now..**

Does anybody has any suspiction or evidence of this?
There is any later documentation about the modem where to check the current definitions for the ES0 and ES1 functions?

**Comment by ralcaide** [comment]
The device is the same but maybe we are a using a different version of the FW. Are you able to get the FW version of both transceivers?

**Comment by juanra-bascones** [comment]
I will look the command manual to find the way to get the fw revision or something alike.

**Comment by ralcaide** [comment]
Yes, it is possible they have a different FW. Please provide here the FW revision and we will ask for the correct documentation

**Comment by juanra-bascones** [comment]
I have collected the fw revision for the HLKRM08 modems in all the boards I have with them.
The result of the "at+RVer" command for them is as follows


Olf-fashioned Hub Board (uart1) 	"at+RVer=V2.5(Mar 22 2017)"
Presence Board:  	                        "at+RVer=V2.5(Mar 22 2017)"  
Hub Board (RF):		                "at+RVer=V2.5(Mar 22 2017)"	
HomeHub Board		                "at+RVer=**V3.06(180227110433)**"
 
that is, all share the same version, but the one in the HomeHub.
